packages/xsnap/test/test-xsnap.js

```diff
@@ -13,44 +13,70 @@ const xsnapOptions = {
   os: os.type(),
 };
 
-test('evaluate and issueCommand', async t => {
+function options() {
   const messages = [];
   async function handleCommand(message) {
     messages.push(decoder.decode(message));
     return new Uint8Array();
   }
-  const vat = xsnap({ ...xsnapOptions, handleCommand });
+  return { ...xsnapOptions, handleCommand, messages };
+}
+
+test('evaluate and issueCommand', async t => {
+  const opts = options();
+  const vat = xsnap(opts);
   await vat.evaluate(`issueCommand(ArrayBuffer.fromString("Hello, World!"));`);
   await vat.close();
-  t.deepEqual(['Hello, World!'], messages);
+  t.deepEqual(['Hello, World!'], opts.messages);
 });
 
 test('evaluate until idle', async t => {
-  const messages = [];
-  async function handleCommand(message) {
-    messages.push(decoder.decode(message));
-    return new Uint8Array();
-  }
-  const vat = xsnap({ ...xsnapOptions, handleCommand });
+  const opts = options();
+  const vat = xsnap(opts);
   await vat.evaluate(`
     (async () => {
       issueCommand(ArrayBuffer.fromString("Hello, World!"));
     })();
   `);
   await vat.close();
-  t.deepEqual(['Hello, World!'], messages);
+  t.deepEqual(['Hello, World!'], opts.messages);
+});
+
+test('idle includes setImmediate too', async t => {
+  const opts = options();
+  const vat = xsnap(opts);
+  await vat.evaluate(`
+    const send = it => issueCommand(ArrayBuffer.fromString(it));
+    setImmediate(() => send("end of crank"));
+    Promise.resolve("turn 2").then(send);
+    send("turn 1");
```

Comment:
I’m suspicious of this technique for measuring end-of-crank in xsnap. I suspect that it’s brittle in the face of client code that manages to schedule more work with another call to `setImmediate`, and unnecessary since we have `fxRunLoop` and can detect idle from the parent side that way.

I think we used `setImmediate` in Node.js for lack of a better option.

Reply:
My understanding, as I wrote in the comment for fc865152e3bc027 , is ...
> In addition to detecting XS Machine quiescense so we can safely take snapshots, the supervisor has to detect vat queiscense so it can tell when a delivery is done.

@warner writes:
> the crank delivery code currently assumes it can get control after the user-level vat code has lost control, but I think the only thing it does with that control is to check the meters, which of course will be done differently on XS

It also reports success / failure of the delivery, yes?

Hmm... I don't seem to have a `catch`, so I guess failure gets dropped on the floor. We don't have a test for that, do we?

Reply:
I have a new plan (#2243), in which liveslots itself is responsible for figuring out when the user-level vat code has lost control (so it can perform some GC work afterwards, and possibly emit GC-related syscalls before end-of-crank). In that world, the Promise returned by `dispatch.deliver()/.notify()`/etc won't fire until end-of-crank, after which the vat ought to be quiescent. The fact that liveslots has access to `setImmediate` means that this might not be true, but the rule is that no syscalls may be made after that Promise fires (i.e. attempting to make a syscall between cranks is vat-fatal, although how exactly that ought to be reported is not clear).

Once that gets implemented, I think we won't need to use `setImmediate` at the top level.